src/Interpreters/SessionTracker.cpp

```diff
@@ -0,0 +1,69 @@
+#include "SessionTracker.h"
+
+#include <Common/NetException.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int USER_SESSION_LIMIT_OVERFLOWN;
+}
+
+SessionInfo::SessionInfo(const ClientInfo & client_info_, const String & session_id_)
+    : client_info(client_info_), session_id(session_id_)
+{
+}
+
+SessionTrackerHandle::SessionTrackerHandle(SessionTracker & tracker_,
+                                           const UUID& user_id_,
+                                           SessionInfoList::const_iterator session_info_iter_) noexcept
+    : tracker(tracker_), user_id(user_id_), session_info_iter(session_info_iter_)
+{
+}
+
+SessionTrackerHandle::~SessionTrackerHandle()
+{
+    tracker.stopTracking(*this);
+}
+
+std::unique_ptr<SessionTrackerHandle>
+SessionTracker::trackSession(const UUID & user_id,
+                             const SessionInfo & session_info,
+                             size_t max_sessions_per_user)
+{
+    std::lock_guard lock(mutex);
+
+    auto user_sessions_iter = sessions_per_user.find(user_id);
+    if (user_sessions_iter == sessions_per_user.end())
+        user_sessions_iter = sessions_per_user.emplace(user_id, SessionInfoList()).first;
+
+    SessionInfoList & session_list = user_sessions_iter->second;
+    if (max_sessions_per_user && session_list.size() >= max_sessions_per_user)
+    {
+        throw NetException(ErrorCodes::USER_SESSION_LIMIT_OVERFLOWN,
+                           "User {} has overflown session count {}",
+                           session_info.client_info.current_user,
+                           max_sessions_per_user);
+    }
+
+    session_list.emplace_front(session_info);
+
+    return std::make_unique<SessionTrackerHandle>(*this, user_id, session_list.begin());
```

Comment:
Yes, I don't like using unique_ptr also, but session and session tracker handle lifetime is not the same. And `session` owns `session tracker handle`. `Session` object can be destroyed even before tracker handle is created (for example in case of authentication fail).

`unique_ptr `protects us from destroying non-created tracker handle.

Actually I can do this:
Add a field to SessionTrackerHandle (is_created).
In case when SessionTracker creates SessionTrackerHandle `is_created `is set to `true` 
When SessionTrackerHandle  is destroyed we could check SessionTrackerHandle::is_created before erasing from the map and set `SessionTrackerHandle::is_created` to false.
This solution will add more complexity than just using `unique_ptr`.

> or (maybe even better) making SessionTrackerHandle an inner (nested) class of SessionTracker and rename it to Session

Ok, Lets do this and ` using SessionTrackerHandle = std::unique_ptr<SessionTracker::Session>`